source/extensions/filters/network/zookeeper_proxy/filter.cc

```diff
@@ -18,56 +20,152 @@ namespace Extensions {
+             "close_resp");
+}
+
+ErrorBudgetResponseType
+ZooKeeperFilterConfig::errorBudgetDecision(const OpCodes opcode,
+                                           const std::chrono::milliseconds latency) const {
+  // Set latency threshold for the current opcode.
+  std::chrono::milliseconds latency_threshold = default_latency_threshold_;
+  int32_t opcode_val = enumToSignedInt(opcode);
+  auto it = latency_threshold_override_map_.find(opcode_val);
+  if (it != latency_threshold_override_map_.end()) {
+    latency_threshold = it->second;
+  }
+
+  // Determine fast/slow response based on the threshold.
+  if (latency <= latency_threshold) {
+    return ErrorBudgetResponseType::FAST;
+  }
+
+  return ErrorBudgetResponseType::SLOW;
 }
 
-void ZooKeeperFilterConfig::initOpCode(OpCodes opcode, Stats::Counter& counter,
-                                       absl::string_view name) {
+void ZooKeeperFilterConfig::initOpCode(OpCodes opcode, Stats::Counter& resp_counter,
+                                       Stats::Counter& resp_fast_counter,
+                                       Stats::Counter& resp_slow_counter, absl::string_view name) {
   OpCodeInfo& opcode_info = op_code_map_[opcode];
-  opcode_info.counter_ = &counter;
+  opcode_info.resp_counter_ = &resp_counter;
+  opcode_info.resp_fast_counter_ = &resp_fast_counter;
+  opcode_info.resp_slow_counter_ = &resp_slow_counter;
   opcode_info.opname_ = std::string(name);
   opcode_info.latency_name_ = stat_name_set_->add(absl::StrCat(name, "_latency"));
 }
 
+int32_t ZooKeeperFilterConfig::getOpCodeIndex(LatencyThresholdOverride_Opcode opcode) {
+  static absl::flat_hash_map<LatencyThresholdOverride_Opcode, int32_t> opcode_map = {
+      {LatencyThresholdOverride::Connect, 0},
+      {LatencyThresholdOverride::Create, 1},
+      {LatencyThresholdOverride::Delete, 2},
+      {LatencyThresholdOverride::Exists, 3},
+      {LatencyThresholdOverride::GetData, 4},
+      {LatencyThresholdOverride::SetData, 5},
+      {LatencyThresholdOverride::GetAcl, 6},
+      {LatencyThresholdOverride::SetAcl, 7},
+      {LatencyThresholdOverride::GetChildren, 8},
+      {LatencyThresholdOverride::Sync, 9},
+      {LatencyThresholdOverride::Ping, 11},
+      {LatencyThresholdOverride::GetChildren2, 12},
+      {LatencyThresholdOverride::Check, 13},
+      {LatencyThresholdOverride::Multi, 14},
+      {LatencyThresholdOverride::Create2, 15},
+      {LatencyThresholdOverride::Reconfig, 16},
+      {LatencyThresholdOverride::CheckWatches, 17},
+      {LatencyThresholdOverride::RemoveWatches, 18},
+      {LatencyThresholdOverride::CreateContainer, 19},
+      {LatencyThresholdOverride::CreateTtl, 21},
+      {LatencyThresholdOverride::Close, -11},
+      {LatencyThresholdOverride::SetAuth, 100},
+      {LatencyThresholdOverride::SetWatches, 101},
+      {LatencyThresholdOverride::GetEphemerals, 103},
+      {LatencyThresholdOverride::GetAllChildrenNumber, 104},
+      {LatencyThresholdOverride::SetWatches2, 105}};
```

Comment:
I may have missed this, but can there be a test that validates what happens when the incoming value is not in the table (the config may be rejected, but just to make sure it is covered).

Reply:
Yeah, there is a [config test for undefined opcode](https://github.com/envoyproxy/envoy/pull/26594/files#diff-b680ac531c7b3eb41602f7d86e5d6bb8271920609161c52bf80eaa7898603913R46-R58).